p2p/transport/webrtc/connection.go

```diff
@@ -0,0 +1,249 @@
+		scope:     scope,
+
+		localPeer:      localPeer,
+		privKey:        privKey,
+		localMultiaddr: localMultiaddr,
+
+		remotePeer:      remotePeer,
+		remoteKey:       remoteKey,
+		remoteMultiaddr: remoteMultiaddr,
+		ctx:             ctx,
+		cancel:          cancel,
+		streams:         make(map[uint16]*dataChannel),
+
+		accept: accept,
+	}
+
+	pc.OnDataChannel(func(dc *webrtc.DataChannel) {
+		log.Debugf("[%s] incoming datachannel: %s", localPeer, dc.Label())
+		id := *dc.ID()
+		var stream *dataChannel
+		dc.OnOpen(func() {
+			// datachannel cannot be detached before opening
+			rwc, err := dc.Detach()
+			if err != nil {
+				log.Errorf("[%s] could not detch channel: %s", localPeer, dc.Label())
+				return
+			}
+			stream = newDataChannel(dc, rwc, pc, nil, nil)
+			conn.addStream(id, stream)
+			accept <- stream
+		})
+
+		dc.OnClose(func() {
+			stream.remoteClosed()
+			conn.removeStream(id)
+		})
+	})
+
+	pc.OnConnectionStateChange(func(state webrtc.PeerConnectionState) {
+		if state == webrtc.PeerConnectionStateClosed || state == webrtc.PeerConnectionStateDisconnected {
+			conn.Close()
+		}
+	})
+
+	return conn
+}
+
+// ConnState implements transport.CapableConn
+func (c *connection) ConnState() network.ConnectionState {
+	return network.ConnectionState{}
+}
+
+// Implement network.MuxedConn
+
+func (c *connection) Close() error {
+	if c.IsClosed() {
+		return nil
+	}
+
+	c.scope.Done()
+	// cleanup routine
+	for _, stream := range c.streams {
+		_ = stream.Close()
+	}
+	c.cancel()
+	_ = c.pc.Close()
+	return nil
+}
+
+func (c *connection) IsClosed() bool {
+	select {
+	case <-c.ctx.Done():
+		return true
+	default:
+	}
+	return false
+}
+
+func (c *connection) OpenStream(ctx context.Context) (network.MuxedStream, error) {
+	type openStreamResult struct {
+		network.MuxedStream
+		error
+	}
+
+	if c.IsClosed() {
+		return nil, os.ErrClosed
+	}
+
+	result := make(chan openStreamResult)
+	dc, err := c.pc.CreateDataChannel("", nil)
+	if err != nil {
+		return nil, err
+	}
+
+	streamID := *dc.ID()
+	var stream *dataChannel
+	dc.OnOpen(func() {
+		rwc, err := dc.Detach()
+		if err != nil {
+			result <- openStreamResult{
+				nil,
+				errDatachannel("could not detach", err),
+			}
+			return
+		}
+		stream = newDataChannel(dc, rwc, c.pc, nil, nil)
+		c.addStream(streamID, stream)
+		result <- openStreamResult{stream, err}
+	})
+
+	dc.OnClose(func() {
+		stream.remoteClosed()
+		c.removeStream(streamID)
+	})
+
+	select {
+	case <-ctx.Done():
+		_ = dc.Close()
+		return nil, ctx.Err()
+	case r := <-result:
+		return r.MuxedStream, r.error
+	}
+}
+
+func (c *connection) AcceptStream() (network.MuxedStream, error) {
+	select {
+	case <-c.ctx.Done():
+		return nil, os.ErrClosed
+	case stream := <-c.accept:
+		return stream, nil
+	}
+}
+
+// implement network.ConnSecurity
+func (c *connection) LocalPeer() peer.ID {
+	return c.localPeer
+}
+
+// only used during setup
+func (c *connection) setRemotePeer(id peer.ID) error {
```

Comment:
1. I did have a separate function for setting the public key but I was not aware of this.
2. I do not have a way around this. When listening, the connection needs to be set up (ie. all the callbacks on the peerconnection need to be set up) before the noise handshake occurs. The reason for this is because if the remote side opens a stream before the local can set up the callbacks, it leads to a case where datachannels are created and not surfaced to the user. We can only learn the remote's peer ID and public key after the noise handshake completes. Therefore we need to initialize a connection and then set it's public key and peer ID if the noise handshake succeeds.